crates/credentials/src/vc.rs

```diff
@@ -127,25 +195,163 @@ impl VerifiableCredential {
+            if vc_exp == i64::default() {
+                vc.expiration_date = Option::from(exp);
+            } else if exp != vc_exp {
+                return Err(CredentialError::ExpirationDateMismatch);
+            }
+
+            let now = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs() as i64;
+
+            if now > vc.expiration_date.unwrap() {
+                return Err(CredentialError::CredentialExpired(
+                    "The verifiable credential has expired".to_string(),
+                ));
+            }
+        } else if registered_claims.expiration.is_none() && vc.expiration_date.is_some() {
+            return Err(CredentialError::MisconfiguredExpirationDate(
+                "VC has expiration date but no exp in registered claims".to_string(),
+            ));
+        }
+
+        // check for consistency and validity of expiration dates
+        if let (Some(exp), Some(vc_exp)) = (registered_claims.expiration, vc.expiration_date) {
+            // Update to registered expiration if current is default (0 or another default if defined)
+            if vc_exp == i64::default() {
+                vc.expiration_date = Some(exp);
+            } else if exp != vc_exp {
+                return Err(CredentialError::ExpirationDateMismatch);
+            }
+
+            // Check if the credential has expired
+            let current_time = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_secs() as i64;
+            if current_time > vc_exp {
+                return Err(CredentialError::CredentialExpired(
+                    "The verifiable credential has expired".to_string(),
+                ));
+            }
+        } else if registered_claims.expiration.is_none() && vc.expiration_date.is_some() {
+            return Err(CredentialError::MisconfiguredExpirationDate(
+                "VC has expiration date but no exp in registered claims".to_string(),
+            ));
+        }
+
+        validate_vc_data_model(&vc).map_err(CredentialError::ValidationError)?;
+
+        Ok(Arc::new(vc))
     }
 
     pub fn decode(jwt: &str) -> Result<Self, CredentialError> {
         let jwt_decoded = Jwt::decode::<VcJwtClaims>(jwt)?;
-
         Ok(jwt_decoded.claims.vc)
     }
 }
 
-// todo we should remove this altogether in the follow-up PR, but it would break bindings so leaving it for now
-pub async fn verify_vcjwt(jwt: &str) -> Result<Arc<VerifiableCredential>, CredentialError> {
-    let jwt_decoded = Jwt::verify::<VcJwtClaims>(jwt).await?;
-    Ok(Arc::new(jwt_decoded.claims.vc))
+fn validate_vc_data_model(
+    vc: &VerifiableCredential,
+) -> Result<(), CredentialDataModelValidationError> {
+    // Required fields
+    if vc.id.is_empty() {
+        return Err(CredentialDataModelValidationError::MissingId);
+    }
+
+    if vc.context.first().map_or(true, |v| v != BASE_CONTEXT) {
```

Comment:
maybe a bit easier on the eyes? I'll defer to your judgement (also holds true for `r#type` below)

```suggestion
    if vc.context.len() == 0 || vc.context[0] != BASE_CONTEXT {
```

Reply:
even better `is_empty()` :)